### DISCUSSION TITLE
UART test software

### DISCUSSION THREAD

**Discussion by lynchaj** [wwarthen/RomWBW#232] [Ideas]
Hi
One idea I've had for a while now is a UART test monitor software.  This would probably be specific to types of RBC computers. Say if it was written for 16C550 style UART it would work with SBC V1, SBC V2, Zeta I, Zeta II, Z80 MBC, etc.  I am thinking an application like FDU or dbgmon to detect a UART, determine what type, print IO port address, if interrupts are enabled, FIFO status, etc.

Tests like detect UART, find UART clock speed, if FIFO buffers are present and/or enabled, various communications parameters settings, speed tests, interrupt tests, DMA tests, etc.  Possibly even a loop-back test with a specially built serial plug (tx -> rx, rts -> cts, dts -> dsr, gnd).  I think this would be useful to determine if UART works and as to what are its capabilities.  For instance 8250, 16450, 16550 all have varying levels of capability although they are backwards compatible.

What are your thoughts on a UART test monitor program?

Thanks, Andrew Lynch

**Comment by wwarthen** [comment]
I personally think something along these lines this is a great idea.

RomWBW itself is far too complex when you just want to confirm the operation of your hardware.  Phil Summers previously suggested a diagnostic ROM that would be built for all of the RomWBW hardware platforms to allow the hardware itself to be validated.

My thought would be to combine these concepts.  Essentially a diagnostic ROM that is like RomWBW itself in that you can just use configuration file settings to build it for any of the supported platforms.  This would hopefully allow for significant code reuse.  When first building or testing a system, you would burn and run the appropriate diagnostic ROM image as a first step to vet the hardware.  The ROM could contain executable modules for functionality like Monitor, Floppy Testing, IDE Testing, etc.

-Wayne

**Comment by lynchaj** [comment]
Hi Wayne

Yes, a diagnostic ROM would be fantastic.  I recall having a diagnostic ROM for IBM PC/XT clones and I was super helpful in diagnosing a flakey computer.  I still have the ROM stashed in my collection of ancient electronic stuff.

I think the initial stub of a UART monitor program already exists buried inside dbgmon.  There is some initialization code there which senses/detects whether the UART is present by probing the UART's scratch register.  It writes a value and then reads it back to indicate whether the UART is present or not.  If not, it halts the CPU otherwise it continues the program.  I think the rest of the UART monitor program would be just adding one UART feature test at a time.

Testing a UART on a one serial port system can be a bit of a trick however you can learn a lot from just the terminal program like minicom.  Also if you have a second serial port or video/keyboard interface you can use a loop-back plug on the serial port to verify which pins are present and how well the UART works.  Loop-back plugs are trivially easy to make so they are not a limiting factor.

Long ago there used to be a PC diagnostic software (PC Tools? Central Point Software?) tool set which included loop-back plugs you could use to verify your serial ports and even the parallel port if I remember correctly.  I had the software and it was actually quite useful in figuring out the weird PC/XT hardware configurations.

From a project stand point the UART monitor seems more achievable than the ROM compression software idea.  Certainly less complicated and can be implemented in phases much more easily.  If/when I start another software project it will probably be this one.

Thanks, Andrew Lynch

**Comment by wwarthen** [comment]
Yes, I think this project is a bit easier to pursue than the ROM disk compression.  It is "standalone" and, as you say, very easy to implement something basic which could then be enhanced over time.  I also see it as adding significant value for initial users.

It should not be hard to layout a simple framework for this.  I cannot work on it right away, but may be able to create some kind of starting point over time.  If someone else wants to take this up, that would be great.

-Wayne
